src/Streamly/Internal/Unicode/Utf8/Reduce.hs

```diff
@@ -0,0 +1,334 @@
+
+-- | 'splitAt' @n t@ returns a pair whose first element is a
+-- prefix of @t@ of length @n@, and whose second is the remainder of
+-- the string. It is equivalent to @('take' n t, 'drop' n t)@.
+--
+-- /Time complexity:/ O(n)
+splitAt :: Int -> Utf8 -> (Utf8, Utf8)
+splitAt = undefined
+
+-- | 'span', applied to a predicate @p@ and text @t@, returns
+-- a pair whose first element is the longest prefix (possibly empty)
+-- of @t@ of elements that satisfy @p@, and whose second is the
+-- remainder of the list.
+--
+-- >> Utf8.span (=='0') "000AB"
+-- ("000","AB")
+--
+-- /Time complexity:/ O(n)
+{-# INLINE span #-}
+span :: (Char -> Bool) -> Utf8 -> (Utf8, Utf8)
+span = undefined
+
+-- | 'break' is like 'span', but the prefix returned is
+-- over elements that fail the predicate @p@.
+--
+-- >> Utf8.break (=='c') "180cm"
+-- ("180","cm")
+--
+-- /Time complexity:/ O(n)
+{-# INLINE break #-}
+break :: (Char -> Bool) -> Utf8 -> (Utf8, Utf8)
+break p = span (not . p)
+
+-- | Group characters in a string according to a predicate.
+--
+-- /Time complexity:/ O(n)
+groupBy :: (Char -> Char -> Bool) -> Utf8 -> [Utf8]
+groupBy p = unsafePerformIO . Stream.toList . Stream.groupsBy p write . toStream
+
+{-
+-- | Returns the /array/ index (in units of 'Word16') at which a
+-- character may be found.  This is /not/ the same as the logical
+-- index returned by e.g. 'findIndex'.
+findAIndexOrEnd :: (Char -> Bool) -> Utf8 -> Int
+findAIndexOrEnd q t@(Utf8 _arr _off len) = go 0
+    where go !i | i >= len || q c       = i
+                | otherwise             = go (i+d)
+                where Iter c d          = iter t i
+-}
+
+-- | Group characters in a string by equality.
+--
+-- /Time complexity:/ O(n)
+group :: Utf8 -> [Utf8]
+group = groupBy (==)
+
+-- | Return all initial segments of the given 'Utf8', shortest
+-- first.
+--
+-- /Time complexity:/ O(n)
+inits :: Utf8 -> [Utf8]
+inits = undefined
+
+-- | Return all final segments of the given 'Utf8', longest
+-- first.
+--
+-- /Time complexity:/ O(n)
+tails :: Utf8 -> [Utf8]
+tails = undefined
+
+-- $split
+--
+-- Splitting functions in this library do not perform character-wise
+-- copies to create substrings; they just construct new 'Utf8's that
+-- are slices of the original.
+
+-- | Break a 'Utf8' into pieces separated by the first 'Utf8'
+-- argument (which cannot be empty), consuming the delimiter. An empty
+-- delimiter is invalid, and will cause an error to be raised.
+--
+-- Examples:
+--
+-- >> splitOn "\r\n" "a\r\nb\r\nd\r\ne"
+-- ["a","b","d","e"]
+--
+-- >> splitOn "aaa"  "aaaXaaaXaaaXaaa"
+-- ["","X","X","X",""]
+--
+-- >> splitOn "x"    "x"
+-- ["",""]
+--
+-- and
+--
+-- > intercalate s . splitOn s         == id
+-- > splitOn (singleton c)             == split (==c)
+--
+-- (Note: the string @s@ to split on above cannot be empty.)
+--
+-- In (unlikely) bad cases, this function's time complexity degrades
+-- towards /O(n*m)/.
```

Comment:
??

Reply:
Now that I read it again, this seems like a weird statement. It should be `O(n*m)`.